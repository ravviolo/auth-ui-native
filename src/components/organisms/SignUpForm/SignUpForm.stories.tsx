import { action } from '@storybook/addon-actions';
import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { View } from 'react-native';
import { StoryScreen } from 'storybook/StoryScreen';

import { SignUpForm } from './SignUpForm';

storiesOf('SignUpForm', module)
  .addDecorator((getStory) => (
    <StoryScreen>
      <View style={{ flex: 0.23 }}>{getStory()}</View>
    </StoryScreen>
  ))
  .add('Default', () => (
    <SignUpForm
      testID={text('testID', 'signup-form-test-id')}
      onSubmitSignUpForm={action('onSubmitSignUpForm')}
    />
  ));
