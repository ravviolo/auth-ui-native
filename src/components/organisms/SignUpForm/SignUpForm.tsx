import { Button, Flex, TextInputHandle } from 'components/atoms';
import { EmailInput, PasswordInput } from 'components/molecules';
import { useEffect, useRef } from 'react';

import { styles } from './SignUpForm.styles';

export interface Props {
  onSubmitSignUpForm: (email: string, password: string) => void;
  testID: string;
}

export const SignUpForm = ({ onSubmitSignUpForm, testID }: Props) => {
  const emailRef = useRef<TextInputHandle | null>(null);
  const passwordRef = useRef<TextInputHandle | null>(null);

  useEffect(() => {
    emailRef.current?.setFocus();
  }, []);

  // todo: console.log('signup render');

  const onSubmit = () => {
    const email = emailRef.current?.getValue();
    const password = passwordRef.current?.getValue();

    if (email && password) {
      onSubmitSignUpForm(email, password);
      emailRef.current?.clear();
      passwordRef.current?.clear();
    }
  };
  return (
    <Flex direction="column" style={styles.container} testID={testID}>
      <EmailInput ref={emailRef} testID="email-input-signup-form-test-id" />
      <PasswordInput ref={passwordRef} testID="password-input-signup-form-test-id" />
      <Button testID="btn-submit-signup-test-id" title="Sign Up" onPress={onSubmit} />
    </Flex>
  );
};
