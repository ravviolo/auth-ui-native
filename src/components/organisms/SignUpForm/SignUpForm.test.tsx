import { render, screen, fireEvent, within } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { SignUpForm } from './SignUpForm';

import type { Props } from './SignUpForm';

describe('SignUpForm', () => {
  const mockOnSubmitSignUpForm = jest.fn<Props['onSubmitSignUpForm'], [string, string]>();

  const testEmail = 'test@test.com';
  const testPassword = 'testpassword';

  const props: Props = {
    testID: 'signup-form-test-id',
    onSubmitSignUpForm: mockOnSubmitSignUpForm,
  };

  it('should render email and password inputs', () => {
    render(<SignUpForm {...props} />);

    const signupForm = screen.getByTestId('signup-form-test-id');
    const emailInput = within(signupForm).getByTestId('email-input-signup-form-test-id');
    const passwordInput = within(signupForm).getByTestId('password-input-signup-form-test-id');

    expect(emailInput).not.toBeNull();
    expect(passwordInput).not.toBeNull();
  });

  it("should render submit button with 'Sign Up' text", () => {
    render(<SignUpForm {...props} />);

    const signupForm = screen.getByTestId('signup-form-test-id');
    const submitButton = within(signupForm).getByTestId('btn-submit-signup-test-id');

    expect(submitButton).toHaveTextContent('Sign Up');
  });

  it('should submit form with non-empty input fields - run submit handler once', () => {
    render(<SignUpForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-signup-test-id');
    const emailInputField = within(
      screen.getByTestId('email-input-signup-form-test-id')
    ).getByTestId('input-field-email-test-id');
    const passwordInputField = within(
      screen.getByTestId('password-input-signup-form-test-id')
    ).getByTestId('input-field-password-test-id');

    fireEvent.changeText(emailInputField, testEmail);
    fireEvent.changeText(passwordInputField, testPassword);
    fireEvent.press(submitButton);

    expect(mockOnSubmitSignUpForm).toHaveBeenCalledTimes(1);
    expect(mockOnSubmitSignUpForm).toHaveBeenCalledWith(testEmail, testPassword);
  });

  it("shouldn't submit form with one or more empty input fields", () => {
    render(<SignUpForm {...props} />);

    const submitButton = screen.getByTestId('btn-submit-signup-test-id');

    fireEvent.press(submitButton);

    expect(mockOnSubmitSignUpForm).not.toHaveBeenCalled();

    const emailInputField = within(
      screen.getByTestId('email-input-signup-form-test-id')
    ).getByTestId('input-field-email-test-id');

    fireEvent.changeText(emailInputField, testEmail);
    fireEvent.press(submitButton);

    expect(mockOnSubmitSignUpForm).not.toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const tree = testRenderer.create(<SignUpForm {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
