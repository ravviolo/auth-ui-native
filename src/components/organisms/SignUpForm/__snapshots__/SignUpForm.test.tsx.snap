// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "column",
      },
      Object {
        "flex": 1,
        "justifyContent": "space-between",
      },
    ]
  }
  testID="signup-form-test-id"
>
  <View
    testID="email-input-signup-form-test-id"
  >
    <Text
      style={
        Object {
          "color": "#3D3D3D",
          "marginBottom": 3,
        }
      }
      testID="input-label-email-test-id"
    >
      Email
    </Text>
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      secureTextEntry={false}
      style={
        Array [
          Object {
            "borderColor": "#9E9E9E",
            "borderRadius": 8,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          },
          Object {
            "borderColor": "#9E9E9E",
          },
        ]
      }
      testID="input-field-email-test-id"
      value=""
    />
  </View>
  <View
    testID="password-input-signup-form-test-id"
  >
    <Text
      style={
        Object {
          "color": "#3D3D3D",
          "marginBottom": 3,
        }
      }
      testID="input-label-password-test-id"
    >
      Password
    </Text>
    <TextInput
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      secureTextEntry={true}
      style={
        Array [
          Object {
            "borderColor": "#9E9E9E",
            "borderRadius": 8,
            "borderWidth": 1,
            "height": 40,
            "padding": 10,
          },
          Object {
            "borderColor": "#9E9E9E",
          },
        ]
      }
      testID="input-field-password-test-id"
      value=""
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FF4684",
        "borderRadius": 10,
        "elevation": 5,
        "padding": 15,
      }
    }
    testID="btn-submit-signup-test-id"
  >
    <Text
      style={
        Object {
          "color": "#EDEDED",
          "letterSpacing": 1,
          "textTransform": "uppercase",
        }
      }
    >
      Sign Up
    </Text>
  </View>
</View>
`;
